### PR TITLE
Changed TokenController._createEthersContract to no longer be async

### PR DESCRIPTION
**Pull request by BelfordZ**
**TokensController touch ups**

The main change in this PR is removal of `async` from the TokensController._createEthersContract interface, since it does not make any async calls, it does not need to return a promise. 

**Description**

- BREAKING:
  - _createEthersContract, though prefixed with an underscore, is very much a public function that could be used by others. That said, no longer returning a promise could be considered a breaking change.

- FIXED:
  - Clean up use of sinon
    - keep track of stubs and directly restore vs. using global `sinon.restore` and hoping for the best.
    - prefer jest mocks over sinon in select places
    - add better typing to TokensController._createEthersContract
    - restore clock / sinon.useFakeTimers manually for to maintain sanity

- REMOVED:
  - remove unnecessary `await` (dont make sync methods that return promises):
    1. call to [tokenStateChangeListener](https://github.com/MetaMask/controllers/pull/895/files#diff-35bf4fcbeee8c75b039b0098c4abab255947963772d6fa3e6ab05823039ea2b1R241)
    2. interface for [TokensController._createEthersContract](https://github.com/MetaMask/controllers/pull/895/files#diff-49200b6dedba3600208082eb048fcb111f5898b373c650b92c966b61409d60cfR550).
  3. call to [tokensController._getNewAllTokensState](https://github.com/MetaMask/controllers/pull/895/files#diff-bf12a1c391e39d40c8a50481d706aaa7ef39c2967197531c7d6e07f9ba9bc67fR826) from tests.
 

- ADDED:
  - create mock / test helper for TokenController (for now only mocking createEthersContract).
    - More DRY testing code, easy to build upon
   

**Checklist**

- [ x ] Tests are included if applicable
- [ x ] Any added code is fully documented

### CODE REVIEW

File: tests/mocks/stubCreateEthers.ts [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import { TokensController } from '../../src/assets/TokensController';
+
+const stubCreateEthers = (ctrl: TokensController, res: boolean) => {
```
Comment (mcmire) [inline tests/mocks/stubCreateEthers.ts:3]: Instead of moving this out into a separate file, what do you think about having inside of the test file itself? In my opinion, when parts of tests are broken up across multiple files, unless there is a good reason for it — say, we have a complicated setup and it would bloat the test file to have everything in one file, or we need to use that setup across multiple files — it makes it difficult to understand the totality of a single test. In this case, I feel like it would be just as readable to have this directly in the test that uses it. What do you think?

Also, although we do prefer Jest mocks instead of Sinon, what do you think about continuing to use Sinon for this file until we make that conversion? Otherwise it might create confusion for someone reading this file.

File: src/assets/TokensController.test.ts [hunk 2]
```diff
@@ -27,13 +29,13 @@ describe('TokensController', () => {
...
 
-    sinon
+    instEthProvStub = sinon
       .stub(tokensController, '_instantiateNewEthersProvider')
       .callsFake(() => null);
   });
 
   afterEach(() => {
-    sinon.restore();
+    instEthProvStub.restore();
```
Comment (mcmire) [inline src/assets/TokensController.test.ts:38]: What do you think about continuing to use Sinon for mocks and keeping `sinon.restore()` here?

File: src/assets/TokensController.test.ts [hunk 3]
```diff
@@ -16,6 +17,7 @@ describe('TokensController', () => {
   let preferences: PreferencesController;
   let network: NetworkController;
 
+  let instEthProvStub: any;
```
Comment (mcmire) [inline src/assets/TokensController.test.ts:20]: Does Sinon give us a type we can use? I feel like we should avoid the use of `any` as much as possible, even in tests.

**Review by mcmire — COMMENTED** [review]
Good change — just had a few suggestions.

File: src/assets/TokensController.test.ts (hunk 2 above)
Comment (BelfordZ) [reply src/assets/TokensController.test.ts:38]: my 2c:

1. we are already using jest, which comes with stub/mocks. Using the mocks that come with the testing framework being used is what most developers would expect.
2. our [jest config uses clear mocks](https://github.com/MetaMask/controllers/blob/main/jest.config.js#L26), which works great when using jest mocks
3. jest has wider support, making very tricky to test situations a lot easier via module-level mocking.
4. we are already have a jest  __mocks__ which indicated to me that we are using jest mocks. intermixing of sinon and jest mocks is not super fun. 

I do like sinon, always have. But when using jest as a testing framework, using sinon just feels a bit stubborn. 

TLDR; We want to use sinon only. I'd argue we should use jest only. Happy to disagree and commit to using sinon.

File: tests/mocks/stubCreateEthers.ts (hunk 1 above)
Comment (BelfordZ) [reply tests/mocks/stubCreateEthers.ts:3]: initially I had this helper fn in the test directly, but I needed to [use it in another test](https://github.com/MetaMask/controllers/pull/895/files#diff-35bf4fcbeee8c75b039b0098c4abab255947963772d6fa3e6ab05823039ea2b1R201) so I exported it, which gave a linting error about no exports from tests. 

Thats why I moved it into a seperate file. The alternative would be to dupe the code around in multiple tests which seems worse.

That said, if we prefer to dupe it, ill make the change.

File: src/assets/TokensController.test.ts (hunk 2 above)
Comment (BelfordZ) [reply src/assets/TokensController.test.ts:39]: Pushed a commit that reverts to using sinon / sinon.restore

Comment (Gudahtt) [reply src/assets/TokensController.test.ts:39]: Related: https://github.com/MetaMask/controllers/issues/744
I'm all for keeping PRs simple though, so changing that separately might be preferable anyway.

**Comment by BelfordZ** [comment]
Lmk if you want me to squash

File: tests/mocks/stubCreateEthers.ts [hunk 4]
```diff
@@ -0,0 +1,12 @@
+import sinon from 'sinon';
+import { TokensController } from '../../src/assets/TokensController';
+
+const stubCreateEthers = (ctrl: TokensController, res: boolean) => {
+  return sinon.stub(ctrl, '_createEthersContract').callsFake(() => {
+    return {
+      supportsInterface: sinon.stub().returns(res),
+    } as any;
```
Comment (BelfordZ) [inline tests/mocks/stubCreateEthers.ts:8]: Wish I could do better than this, but ethers.Contract makes the contract read functions as `readonly` and that makes it really hard to return anything here that will satisfy the full interface.

File: tests/mocks/stubCreateEthers.ts (hunk 1 above)
Comment (mcmire) [reply tests/mocks/stubCreateEthers.ts:4]: Yeah, if this were an implementation file then I would say it'd be fine to extract, but I think a little duplication is okay in test files, especially if it is 2 lines of code.

File: src/assets/TokensController.test.ts (hunk 2 above)
Comment (mcmire) [reply src/assets/TokensController.test.ts:39]: @BelfordZ Okay thanks. As Mark pointed out moving everything to Jest mocks in this repo is something we definitely we want to do, and I agree with all your points about using Jest over Sinon. The thing I forgot to say here is that I think from a maintainability perspective, it would be better to convert all tests in a file to use Jest mocks in one go rather than having some tests use Jest mocks and some use Sinon mocks. The current situation is definitely confusing but I feel like that would help control the chaos slightly better. So that's where my suggestion came from. That said if you feel like taking on converting a file we would definitely appreciate it :)

File: src/assets/TokensController.ts [hunk 5]
```diff
@@ -547,16 +547,12 @@ export class TokensController extends BaseController<
     }
   }
 
-  async _createEthersContract(
+  _createEthersContract(
```
Comment (mcmire) [inline src/assets/TokensController.ts:550]: I just want to say, I don't agree with what we are doing here. This method should not exist, and I don't think we should be stubbing `ethers.Contract`. I think it would be better to use dependency injection and have TokensController take a Contract constructor which can be swapped out at runtime for something else. That would obviate the need for this method. I realize we can't really change this right now given that there is a test in the extension which relies on stubbing this method, but I think that would be the next logical move here.

Comment (BelfordZ) [reply src/assets/TokensController.ts:550]: agreed, the whole thing is anti-patterns on anti-patterns. But its okay. At this point, anything that helps towards resolving test non-determinism is good. The next refactor will be a little bit smaller and easier. Many more small improvements will make the big improvement PRs much less busy.

Comment (mcmire) [reply src/assets/TokensController.ts:550]: Awesome, look forward to it!

**Review by mcmire — APPROVED** [review]
Looks good to me!
